2067 contact trace frontend

clemens-msupply opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Part of #2019

The contact trace UI works very similar to the encounter UI with the difference that a contact trace is created in the DetailView (encounter is created when scheduled and then edited in the DetailView).

Currently, a contact trace can be created without being enrolled in the matching program. For example, a contact trace might be created for a patient enrolled in a Testing OR a Care program, for this reason there is no restriction right now when a contact trace can be created (@richardmoizeau).

- Adds an "Add Contact Trace" option to the add button in the patient view
- This button leads to a modal that allows you to select a tracing type 
- The modal leads to an DetailView that can be filled to create the contact trace 
- The same DetailView is used to edit a contact trace
- There is a new Contact Traces tab with a list of contact traces for a single patient
- There is also a list in the left sidebar showing all contact traces for all patients

![image](https://github.com/openmsupply/open-msupply/assets/88299195/3fabf15d-88d2-4a04-b920-d175a5258538)

# 🧪 How has/should this change been tested? 
<!-- Explain how to setup for testing here if it is not already obvious, and how you've tested this PR. -->

Please use the latest dev init script from the schema repository.
- Create a contact trace
- Edit a contact trace
- Check both contact trace lists

mark-prins commented on client/packages/system/src/ContactTrace/DetailView/DetailView.tsx:
```diff
@@ -0,0 +1,270 @@
+        </span>
+      );
+    }
+  }, [contactData]);
```
🤔  the hard coded ` / ` is troubling me a little.  I wonder if a utility method in the breadcrumb files would be useful?

with 
```
  const { appendBreadcrumb, setSuffix } = useBreadcrumbs([
    AppRoute.ContactTrace,
  ]);
```

and the update in useBreadcrumbs

```
type BreadcrumbState = {
  appendBreadcrumb: (element: ReactElement) => void;
  setSuffix: (suffix?: string | ReactElement) => void;
  suffix?: (string | ReactElement)[];
  setUrlParts: (urlParts: UrlPart[]) => void;
  urlParts: UrlPart[];
};

const useBreadcrumbState = create<BreadcrumbState>(set => ({
  setSuffix: suffix =>
    set(state => ({ ...state, suffix: suffix ? [suffix] : undefined })),
  appendBreadcrumb: element =>
    set(state => {
      const suffix = state.suffix ? [...state.suffix, element] : [element];
      return {
        ...state,
        suffix,
      };
    }),
  suffix: undefined,
  setUrlParts: (urlParts: UrlPart[]) => set(state => ({ ...state, urlParts })),
  urlParts: [],
}));
```

you can
```
  useEffect(() => {
    if (contactData) {
      const LinkToPatient = (
        <Breadcrumb
          to={RouteBuilder.create(AppRoute.Dispensary)
            .addPart(AppRoute.Patients)
            .addPart(contactData.patient.id)
            .addQuery({ tab: PatientTabValue.ContactTraces })
            .build()}
        >
          {getLocalisedFullName(
            contactData.patient?.firstName,
            contactData.patient?.lastName
          )}
        </Breadcrumb>
      );
      const ProgramDetails = (
        <span>{`${contactData.programName} - ${dateFormat.localisedDate(
          contactData.contactTrace.datetime
        )}`}</span>
      );

      setSuffix(LinkToPatient);
      appendBreadcrumb(ProgramDetails);
    }
  }, [contactData]);
```

clemens-msupply replied:
what do you mean with hard coded `/`? You mean that last program details parts looks like its another step?

Can't we just move the ProgramDetails out of the breadcrumb? e.g. right align it?

`appendBreadcrumb` is like `setPostfix` right? you can only append one more part?

mark-prins replied:
> what do you mean with hard coded /? You mean that the last program details parts looks like its another step?

No, just that there is a `/` as a string hard coded in this sort of line

```
<span>{` / ${contactData.programName} - ${dateFormat.localisedDate(
            contactData.contactTrace.datetime
          )}`}</span>
```

and that feel inelegant. The rest of the breadcrumb is an array of items, and the breadcrumb component decides how to format and display the parts. Now, with the code above, we are appending a big string and essentially overriding the function of the breadcrumb display component.

> Can't we just move the ProgramDetails out of the breadcrumb? e.g. right align it?

Could do, I guess. The original intention of the `setSuffix` was to have control over the formatting of the last portion of the breadcrumb, and I think that's still valid here. It's just that we're now adding two 'last portions' which is stretching the intention somewhat.

> appendBreadcrumb is like setPostfix right? you can only append one more part?

not really - the `appendBreadcrumb` is adding another item to the array, whereas `setSuffix` is setting the final element of the array. I originally had `appendBreadcrumb` called twice, before realising that the DetailView component is re-rendering and therefore appends twice.. which is why I'm calling `setSuffix` then `appendBreadcrumb` to add another part

mark-prins replied:
another option would be to provide a formatter function to the breadcrumb component, which accepts something as a parameter and formats the final part of the breadcrumb. just thinking out loud, I have no intention to try that out.

clemens-msupply replied:
Or maybe `setSuffixes` so that you can set multiple suffixes if needed? Each part would be separated by a `/`.

So there would be the urlParts from the current location and then a list of additional suffix parts. A suffix part could be, for example, a <Breadcrumb> or other components.

mark-prins replied:
yes, that would work too - I wouldn't make changes now though, best to make a new issue

clemens-msupply replied:
https://github.com/openmsupply/open-msupply/issues/2118

mark-prins commented on client/packages/system/src/Encounter/ListView/columns.ts:
```diff
@@ -17,7 +17,7 @@ import { useLogicalStatus } from '../utils';
 
 interface useEncounterListColumnsProps {
-  onChangeSortBy: (column: Column<any>) => void;
+  onChangeSortBy: (column: Column<EncounterRowFragment>) => void;
```
nice!

mark-prins commented on client/packages/system/src/Patient/api/hooks/document/usePatient.ts:
```diff
@@ -1,7 +1,7 @@
 import { usePatientApi } from '../utils/usePatientApi';
 
-export const usePatient = (nameId: string) => {
+export const usePatient = (nameId: string | undefined) => {
```
```suggestion
export const usePatient = (nameId?: string) => {
```

clemens-msupply replied:
This was intentionally. It doesn't make sense to call usePatient without parameter but the user might provide a string that is undefined, e.g. because nameId is not know yet. This syntacx enforces that a parameter must be provided.

mark-prins replied:
ah, right.

mark-prins reviewed (changes requested):
🎉  that's a comprehensive PR! nicely done - overall that's great, I just have a few concerns and suggestions, as commented.

a last minute thought - how about showing rows in the list view as greyed-out if the status is `Done`? What about disabling editing if the status is `Done`? 

I find the use of the same icon confusing in the contact tracing detail view. My vote would be to use `document` and `medical-chart` for `Contact Details` and `HIV Tracing` respectfully

<img width="545" alt="Screenshot 2023-08-15 at 5 29 28 PM" src="https://github.com/openmsupply/open-msupply/assets/9192912/1938def7-bee7-46a2-b07e-9aab756804c1">

clemens-msupply commented on client/packages/common/src/intl/locales/en/app.json:
```diff
@@ -92,6 +92,7 @@
   "patients": "Patients",
   "dispensary": "Dispensary",
   "programs": "Programs",
+  "contact-trace": "Contact Traces",
```
Any suggestion how to name it internally? e.g. ContactTracingEntry?

mark-prins replied:
I think `ContactTrace` is ok internally, it can be a noun and so within the right context it's ok.

clemens-msupply commented on client/packages/system/src/ContactTrace/ListView/columns.ts:
```diff
@@ -0,0 +1,71 @@
+  const { localisedDate } = useFormatDateTime();
+
+  const columnList: ColumnDescription<ContactTraceRowFragment>[] = [];
+  if (includePatient) {
+    columnList.push({
+      key: 'patientName',
+      label: 'label.patient',
+      accessor: ({ rowData }) => rowData?.patient?.name,
+    });
+  }
+  columnList.push(
+    {
+      key: 'programName',
+      label: 'label.program',
+      accessor: ({ rowData }) => rowData.program.name,
+    },
+    {
+      key: 'contactTraceId',
+      label: 'label.contact_trace-id',
+    },
+    {
+      key: 'contactPatientName',
+      label: 'label.contact-patient',
+      accessor: ({ rowData }) => rowData?.contactPatient?.name,
+    },
+    {
+      key: 'status',
+      label: 'label.status',
+    },
+    {
+      key: 'datetime',
+      label: 'label.date',
+      align: ColumnAlign.Right,
+      width: 175,
+      formatter: dateString =>
+        dateString ? localisedDate((dateString as string) || '') : '',
+    }
+  );
```
thanks! didn't thought about trying this out...

clemens-msupply commented on client/packages/system/src/Patient/ContactTrace/Components/ListView.tsx:
```diff
@@ -0,0 +1,77 @@
+import { useContactTraces } from '@openmsupply-client/programs';
+import { usePatient } from '../../api';
+import { createQueryParamsStore, useQueryParamsStore } from '@common/hooks';
+import { ContactTraceRowFragment } from 'packages/programs/src/api/operations.generated';
+import { useContactTraceListColumns } from 'packages/system/src/ContactTrace';
+import { AppRoute } from 'packages/config/src';
```
Sorry about these should have checked

clemens-msupply commented on client/packages/system/src/ContactTrace/DetailView/Toolbar.tsx:
```diff
@@ -0,0 +1,138 @@
+import React, { FC, ReactNode, useEffect, useState } from 'react';
+import {
+  AppBarContentPortal,
+  Box,
+  InputWithLabelRow,
+  Grid,
+  useTranslation,
+  BasicTextInput,
+  DatePickerInput,
+  DateUtils,
+  UserIcon,
+  LocaleKey,
+  TypedTFunction,
+  Option,
+  ContactTraceNodeStatus,
+} from '@openmsupply-client/common';
+
+import { Select } from '@common/components';
```
Auto import I guess ...

clemens-msupply commented on client/packages/system/src/ContactTrace/DetailView/Footer.tsx:
```diff
@@ -0,0 +1,152 @@
+  trace,
+}) => {
+  const t = useTranslation('common');
+  // TODO StatusCrumbs shows "Order History" in the logs pop up -> make this configurable
```
This is mostly copy and past from the encounter view...

clemens-msupply commented on client/packages/system/src/ContactTrace/DetailView/DetailView.tsx:
```diff
@@ -0,0 +1,270 @@
+    }
+
+    const contactTrace = contactTraces?.nodes?.[0];
+    const data: ContactTraceData | undefined = contactTrace
```
I am tending to leave it in. The reason I am doing this is that durring development the type errors get very nasty if you have some simple errors like a missing type. When providing the target type you get much better help from the compiler....

clemens-msupply commented on client/packages/system/src/ContactTrace/DetailView/DetailView.tsx:
```diff
@@ -0,0 +1,270 @@
+  traceId: string,
+  createType: string | null,
+  createPatientId: string | null
+): { data: ContactTraceData | undefined; isLoading: boolean } => {
```
Yes but I find this much clearer since the type is not trivial. Furthermore, I like to have this as a contract to enforce what needs to be returned (see other comment).

clemens-msupply commented on client/packages/system/src/ContactTrace/DetailView/DetailView.tsx:
```diff
@@ -0,0 +1,270 @@
+  createPatientId: string | null
+): { data: ContactTraceData | undefined; isLoading: boolean } => {
+  const [result, setResult] = useState<{
+    data: ContactTraceData | undefined;
```
Similar to previous comment, I want to enforce that a data field is returned in all code paths, e.g. so that I don't forget to return the data in some cases.

clemens-msupply commented on client/packages/system/src/ContactTrace/DetailView/DetailView.tsx:
```diff
@@ -0,0 +1,270 @@
+  useEffect(() => {
+    if (contactData) {
+      setSuffix(
+        <span key="patient-contact-trace">
```
You get a `Warning: Each child in a list should have a unique "key" prop.` when there is no key...

clemens-msupply commented on client/packages/programs/src/api/hooks/document/useUpsertContactTrace.ts:
```diff
@@ -0,0 +1,60 @@
+ * @param onUpsert if provided this callback will be called whenever a contact trace has been
+ * updated
+ */
+export const useUpsertContactTrace = (
```
I removed this hook. It was used early durring the development but is not needed right now.

clemens-msupply commented on client/packages/system/src/ContactTrace/DetailView/DetailView.tsx:
```diff
@@ -0,0 +1,270 @@
+    createPatientId
+  );
+  // Used when creating a new contact trace:
+  const [newContactTraceId, setNewContactTraceId] = useState(id);
```
done thanks

clemens-msupply commented on client/packages/system/src/ContactTrace/DetailView/DetailView.tsx:
```diff
@@ -0,0 +1,270 @@
+import { PatientTabValue } from '../../Patient/PatientView/PatientView';
+import { usePatient } from '../../Patient';
+
+export type ContactTrace = {
```
I moved the useContactTraceData hook into its own file since the data types belongs to it and it would be hard to understand what the types are used for otherwise...

clemens-msupply commented on client/packages/programs/src/api/api.ts:
```diff
@@ -401,3 +407,59 @@ export const getProgramEventQueries = (sdk: Sdk, storeId: string) => ({
+};
+
+export const getContactTraceQueries = (sdk: Sdk, storeId: string) => ({
+  contactTraces: async ({
```
I renamed them. Find it confusing to name them like this though. For example, its usually used like:

```ts
  const api = useContactTraceApi();
  return useMutation(api.insert, {
```

The api endpoint is `insertContactTrace` though and not insert... Find the api folder quite hard to read/write since you need to go through so many files to get to the part that actually calls the api, having the original endpoint name helps a little bit.

mark-prins replied:
sure - up to you then. 
another suggestion is to be more specific with the local var

```
const contactTraceApi = useContactTraceApi();
  return useMutation(contactTraceApi.insert, {
```

but it's fine either way. objection withdrawn :)

mark-prins commented on client/packages/system/src/ContactTrace/DetailView/DetailView.tsx:
```diff
@@ -0,0 +1,270 @@
+            to={RouteBuilder.create(AppRoute.Dispensary)
+              .addPart(AppRoute.Patients)
+              .addPart(contactData.patient.id)
+              .addQuery({ tab: PatientTabValue.ContactTraces })
```
yeah, I'm not very happy with the `useUrlQueryParams` implementation and interaction with the queryParamsStore. The two hooks should probably be combined, with a parameter to specify whether you want to use the URL query params or not. Mostly the api queries should be using url query parameters, to keep the page states bookmarkable.. that doesn't apply though when there are nested lists ( such as lists on a modal ) so we need to be able to opt out of the url query parameters.

mark-prins reviewed (approved):
👍  thanks for the changes!
It sounds like the detail and list views will have to change with Richard's new issue - that aside, I think this PR is good to merge 🙏

clemens-msupply commented:
@mark-prins thanks for the review think I addressed all comment but as you said more changes are to come. Plan to keep the contact tracing (especially backend) in a separate branch till it has settled down.
